v3io/dataplane/transport/httpclient.py

```diff
@@ -102,18 +127,36 @@ def wait_response(self, request, raise_for_status=None, num_retries=1):
 
                 response = v3io.dataplane.response.Response(request.output, status_code, headers, response_body)
 
-                self._free_connections.put(connection, block=True)
+                # Return connection to pool if successful
+                with self._lock:
+                    if not self._closed:
+                        try:
+                            self._free_connections.put(connection, block=False)
+                        except queue.Full:
+                            # If queue is full, close the extra connection
+                            connection.close()
 
                 response.raise_for_status(request.raise_for_status or raise_for_status)
-
                 return response
 
             except v3io.dataplane.response.HttpResponseError as response_error:
                 self._logger.warn_with(f"Response error: {response_error}")
+                # Return connection to pool even on HTTP errors, as the connection is still valid
+                with self._lock:
+                    if not self._closed:
+                        try:
+                            self._free_connections.put(connection, block=False)
+                        except queue.Full:
+                            connection.close()
                 raise response_error
             except BaseException as e:
-                connection.close()
-                connection = self._create_connection(self._host, self._ssl_context)
+                # Handle connection error in thread-safe way
+                with contextlib.suppress(Exception):
+                    connection.close()
+                with self._lock:
+                    if self._closed:
+                        raise RuntimeError("Transport closed during response handling") from e
+                    connection = self._create_connection(self._host, self._ssl_context)
```

Comment:
Doesn't seem like this line belongs in the locked section.

Reply:
It can be and it does not affect the performance. Eventually it calls http.client.HTTPConnection, which should be thread safe, but since all it does is initializing some structures without actually doing any connecting, I prefer to have it under the lock.

Reply:
But why should it be under the lock if it doesn't require locking?